python/dgl/batch.py

```diff
@@ -0,0 +1,399 @@
+"""Utilities for batching/unbatching graphs."""
+
+from collections import defaultdict
+from collections.abc import Mapping
+
+from . import backend as F
+from .base import ALL, is_all, DGLError, dgl_warning
+from . import convert
+from . import utils
+
+__all__ = ['batch', 'unbatch']
+
+def batch(graphs, ndata=ALL, edata=ALL, *, node_attrs=None, edge_attrs=None):
+    r"""Batch a collection of ``DGLGraph``s into one graph for more efficient
+    graph computation.
+    
+    Each input graph becomes one disjoint component of the batched graph. The nodes
+    and edges are relabeled to be disjoint segments:
+
+    =================  =========  =================  ===  =========
+                       graphs[0]  graphs[1]          ...  graphs[k]
+    =================  =========  =================  ===  =========
+    Original node ID   0 ~ N_0    0 ~ N_1            ...  0 ~ N_k
+    New node ID        0 ~ N_0    N_0+1 ~ N_0+N_1+1  ...  1+\sum_{i=0}^{k-1} N_i ~ 1+\sum_{i=0}^k N_i 
+    =================  =========  =================  ===  =========
+    
+    Because of this, many of the computations on a batched graph are the same as if
+    performed on each graph individually, but become much more efficient
+    since they can be parallelized easily. This makes ``dgl.batch`` very useful
+    for tasks dealing with many graph samples such as graph classification tasks.
+
+    For heterograph inputs, they must share the same set of relations (i.e., node types
+    and edge types) and the function will perform batching on each relation one by one.
+    Thus, the result is also a heterograph and has the same set of relations as the inputs.
+    
+    The numbers of nodes and edges of the input graphs are accessible via the
+    :func:`DGLGraph.batch_num_nodes` and :func:`DGLGraph.batch_num_edges` attributes
```

Comment:
I presume `DGLGraph.batch_num_nodes` and `DGLGraph.batch_num_edges` are now functions with node type/edge type introduced?